### PR TITLE
ZCS-9475:- Mobile Sync Admin Console implementation

### PR DESCRIPTION
**Pull request by madhavdhuppe**

### CODE REVIEW

File: WebRoot/css/zmadmin.css [hunk 1]
```diff
@@ -341,6 +341,7 @@ INPUT.admin_xform_name_input_disabled {
 	width:auto;
 }
 
+
```
Comment (BhavinBathani) [inline WebRoot/css/zmadmin.css:344]: Do we need this extra blank line?

File: WebRoot/js/zimbraAdmin/common/ZaItem.js [hunk 2]
```diff
@@ -115,6 +115,32 @@ function(a,b) {
 	return ZaItem.compareAttr(a,b,"description");
 }
 
+ZaItem.compareEmailAddressAsc = 
+function(a,b) {
```
Comment (BhavinBathani) [inline WebRoot/js/zimbraAdmin/common/ZaItem.js:119]: Can we use more proper variable names?

File: WebRoot/js/zimbraAdmin/common/ZaItem.js [hunk 3]
```diff
@@ -115,6 +115,32 @@ function(a,b) {
...
+	var al = a.emailAddress.toLowerCase();
+	var bl = b.emailAddress.toLowerCase();
+
+	if (al < bl)
```
Comment (BhavinBathani) [inline WebRoot/js/zimbraAdmin/common/ZaItem.js:123]: use curly braces to conditional statements

File: WebRoot/js/zimbraAdmin/common/ZaItem.js [hunk 4]
```diff
@@ -115,6 +115,32 @@ function(a,b) {
...
+		return -1;
+	if (al > bl)
+		return 1;
+	else
+		return 0;
+}
+
+ZaItem.compareEmailAddressDesc =
```
Comment (BhavinBathani) [inline WebRoot/js/zimbraAdmin/common/ZaItem.js:131]: can we club this function with above function? lets have 3rd parameter to identify asc or desc.

File: WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js [hunk 5]
```diff
@@ -0,0 +1,207 @@
+/*
+ * ***** BEGIN LICENSE BLOCK *****
+ * Zimbra Collaboration Suite Web Client
+ * Copyright (C) 2012, 2013, 2014, 2016 Synacor, Inc.
```
Comment (BhavinBathani) [inline WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js:4]: Don't we need to add 2020 ?

File: WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    var soapDoc = AjxSoapDoc.create("AllowDeviceRequest",ZaZimbraAdmin.URN, null);
+
+    var device = soapDoc.set("device", null, null);
+    device.setAttribute("id", obj.id);
+    
+    var account = soapDoc.set("account", obj[ZaRegisterDevice.RD_Email_Address]);
+    account.setAttribute("by", "name");
+    
+    // obj.status = 0;
+    // return [obj];
```
Comment (BhavinBathani) [inline WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js:161]: Remove commented code.

File: WebRoot/messages/ZaMsg.properties [hunk 7]
```diff
@@ -2761,6 +2767,27 @@ EditWarningAcceptMsg = Cannot change type of volume {0} because it is the curren
...
+MB_But_Wipe = Wipe Device
+MB_Need_Prov = Needs provisioning
+MB_Active = Active
+MB_Suspended = Suspended
+MB_Wipe_ACK = Wipe pending
+MB_Wipe_Comp = Wipe completed
+MB_Blocked = Blocked
+
+
+
```
Comment (BhavinBathani) [inline WebRoot/messages/ZaMsg.properties:2790]: Do we need this extra blank line?

File: WebRoot/js/zimbraAdmin/rp/view/ZaRetentionPolicyListView.js [hunk 8]
```diff
@@ -34,6 +34,7 @@ ZaRetentionPolicyListView.prototype.toString = function() {
 	return "ZaRetentionPolicyListView";
 };
 
+
```
Comment (BhavinBathani) [inline WebRoot/js/zimbraAdmin/rp/view/ZaRetentionPolicyListView.js:37]: Do we need this extra blank line?

File: WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+        var resp = ZaRequestMgr.invoke(params, reqMgrParams).Body.GetDeviceStatusResponse.device;
+        return resp;
+    } catch(ex) {
+        throw ex;
+        return null;
+    }
+}
+
+
```
Comment (BhavinBathani) [inline WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js:97]: Do we need this extra blank line?

File: WebRoot/js/zimbraAdmin/globalconfig/view/GlobalConfigXFormView.js [hunk 10]
```diff
@@ -388,6 +451,11 @@ GlobalConfigXFormView.AUTO_PROV_TAB_RIGHTS = [];
 GlobalConfigXFormView.RETENTION_POLICY_TAB_ATTRS = [];
 GlobalConfigXFormView.RETENTION_POLICY_TAB_RIGHTS = [];
 
+GlobalConfigXFormView.REGISTERED_DEVICES_TAB_ATTRS = []
+GlobalConfigXFormView.REGISTERED_DEVICES_TAB_RIGHTS = []
+
+
+
```
Comment (BhavinBathani) [inline WebRoot/js/zimbraAdmin/globalconfig/view/GlobalConfigXFormView.js:458]: Do we need this extra blank line?

File: WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+ZaRegisterDevice.quarantineDevice = function(obj) {
+    var soapDoc = AjxSoapDoc.create("QuarantineDeviceRequest",ZaZimbraAdmin.URN, null);
+
+    var device = soapDoc.set("device", null, null);
+    device.setAttribute("id", obj.id);
+    
+    var account = soapDoc.set("account", obj[ZaRegisterDevice.RD_Email_Address]);
+    account.setAttribute("by", "name");
+    // obj.status = 2;
+    // return [obj];
```
Comment (BhavinBathani) [inline WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js:107]: Remove commented code.

File: WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+        throw ex;
+        return null;
+    }
+}
+
+ZaRegisterDevice.wipeDevice = function(obj) {
+    var soapDoc = AjxSoapDoc.create("RemoteWipeRequest",ZaZimbraAdmin.URN, null);
+
+    // obj.status = 3;
+    // return [obj];
```
Comment (BhavinBathani) [inline WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js:184]: Remove commented code.

File: WebRoot/messages/ZaMsg.properties [hunk 13]
```diff
@@ -2378,6 +2378,12 @@ BUSY_GET_RETENTION_POLICIES = Getting retention polices ...
 BUSY_CREATE_RETENTION_POLICIES = Creating retention polices ...
 BUSY_DELETE_RETENTION_POLICIES = Deleting retention polices ...
 BUSY_MODIFY_RETENTION_POLICIES = Modifying retention polices ...
+BUSY_GETTING_SYNC_DEVICES = Getting synced devices ...
```
Comment (BhavinBathani) [inline WebRoot/messages/ZaMsg.properties:2381]: What about translations for other languages?
Create another ticket to track it if you are not going to cover the same in this ticket.

File: WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js (hunk 6 above)
Comment (madhavdhuppe) [reply WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js:161]: removed was added for testing in local vm

File: WebRoot/js/zimbraAdmin/mobileDevices/view/ZaRegisteredDeviceListView.js [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+		for(var i = 0; i < cnt; i++) {
+			var field = this._headerList[i]._field;
+
+			if(field == ZaRegisterDevice.RD_Email_Address) {
+				html[idx++] = "<td align=left height=20px width=" + this._headerList[i]._width + ">";
+				html[idx++] = AjxStringUtil.htmlEncode(item[ZaRegisterDevice.RD_Email_Address]);
+				html[idx++] = "</td>";
+			} else if(field == ZaRegisterDevice.RD_Last_Login) {
+				html[idx++] = "<td align=left height=20px width=" + this._headerList[i]._width + ">";
+                html[idx++] = AjxStringUtil.htmlEncode(ZaRegisterDevice.getDateDifference(item[ZaRegisterDevice.RD_Last_Login]));
```
Comment (BhavinBathani) [inline WebRoot/js/zimbraAdmin/mobileDevices/view/ZaRegisteredDeviceListView.js:61]: Adjust indentation.

File: WebRoot/js/zimbraAdmin/mobileDevices/view/ZaRegisteredDeviceListView.js [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+ZaRegisteredDeviceListView.prototype = new ZaListView;
+ZaRegisteredDeviceListView.prototype.constructor  = ZaRegisteredDeviceListView;
+ZaRegisteredDeviceListView.prototype.toString = function() {
+	return "ZaRegisteredDeviceListView";
+};
+
+ZaRegisteredDeviceListView.prototype._createItemHtml =
+function(item) {
+	var html = new Array(50);
+	var	div = document.createElement("div");
```
Comment (BhavinBathani) [inline WebRoot/js/zimbraAdmin/mobileDevices/view/ZaRegisteredDeviceListView.js:40]: remove tab after var

File: WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+    var params = new Object();
+    params.soapDoc = soapDoc;
+
+    try{
+        var reqMgrParams = {
+            controller : ZaApp.getInstance().getCurrentController(),
+            busyMsg : ZaMsg.BUSY_RESUMING_SYNC_DEVICES
+        };
+
+        var resp = ZaRequestMgr.invoke(params, reqMgrParams).Body.AllowDeviceResponse.device;
```
Comment (BhavinBathani) [inline WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js:172]: no need to have resp variable. return directly

Comment (madhavdhuppe) [reply WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js:172]: done

File: WebRoot/js/zimbraAdmin/mobileDevices/view/ZaRegisteredDeviceListView.js (hunk 14 above)
Comment (madhavdhuppe) [reply WebRoot/js/zimbraAdmin/mobileDevices/view/ZaRegisteredDeviceListView.js:61]: done

File: WebRoot/js/zimbraAdmin/mobileDevices/view/ZaRegisteredDeviceListView.js (hunk 15 above)
Comment (madhavdhuppe) [reply WebRoot/js/zimbraAdmin/mobileDevices/view/ZaRegisteredDeviceListView.js:40]: done

File: WebRoot/js/zimbraAdmin/common/ZaItem.js (hunk 4 above)
Comment (madhavdhuppe) [reply WebRoot/js/zimbraAdmin/common/ZaItem.js:131]: ok

File: WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js (hunk 5 above)
Comment (madhavdhuppe) [reply WebRoot/js/zimbraAdmin/mobileDevices/model/ZaRegisterDevice.js:4]: added 2020

File: WebRoot/js/zimbraAdmin/common/ZaItem.js (hunk 2 above)
Comment (madhavdhuppe) [reply WebRoot/js/zimbraAdmin/common/ZaItem.js]: updated

File: WebRoot/js/zimbraAdmin/common/ZaItem.js (hunk 3 above)
Comment (madhavdhuppe) [reply WebRoot/js/zimbraAdmin/common/ZaItem.js]: updated

File: WebRoot/messages/ZaMsg.properties (hunk 13 above)
Comment (madhavdhuppe) [reply WebRoot/messages/ZaMsg.properties]: Added translations

**Review by BhavinBathani — APPROVED** [review]
LGTM

File: WebRoot/js/zimbraAdmin/common/ZaItem.js [hunk 17]
```diff
@@ -115,6 +115,35 @@ function(a,b) {
...
+		} else {
+
+			if (object1Email < object2Email) {
+				return 1;
+			}
+			
+			if (object1Email > object2Email) {
+				return -1;
+			}
+		}
```
Comment (auminpatel) [inline WebRoot/js/zimbraAdmin/common/ZaItem.js:141]: can we write it like this 
```
if (object1Email < object2Email) {
	return isAsc ? -1 : 1;
}

if (object1Email > object2Email) {
		return isAsc ? 1 : -1;
}
```

Comment (madhavdhuppe) [reply WebRoot/js/zimbraAdmin/common/ZaItem.js]: updated

File: WebRoot/js/zimbraAdmin/common/ZaListView.js [hunk 18]
```diff
@@ -154,10 +154,21 @@ function (ev, listEv, clickedEl) {
...
+		}
 	} else {
-    	this._list.sort(ZaItem.compareNamesDesc);
+
+		if(this.toString() === "ZaRegisteredDeviceListView") {
+			this._list.sort(ZaItem.compareEmailAddress(false));
+		} else {
+			this._list.sort(ZaItem.compareNamesDesc);
+		}
 	}
```
Comment (auminpatel) [inline WebRoot/js/zimbraAdmin/common/ZaListView.js:170]: can we write it 
if(this.toString() === "ZaRegisteredDeviceListView") {
			this._list.sort(ZaItem.compareEmailAddress(true));
		} else {
			bSortAsc  ? this._list.sort(ZaItem.compareNamesAsc) : this._list.sort(ZaItem.compareNamesDesc);
		}

Comment (madhavdhuppe) [reply WebRoot/js/zimbraAdmin/common/ZaListView.js:170]: we cant write like that ZaItem.compareEmailAddress(false) is also there

Comment (auminpatel) [reply WebRoot/js/zimbraAdmin/common/ZaListView.js:170]: My bad dint see the false

**Review by auminpatel — APPROVED** [review]
LGTM

**Review by BhavinBathani — APPROVED** [review]

**Pull request merged**